core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+        obj = mock("object")
+        obj.should_receive(:inspect).and_return("<inspect-result>")
+        format("%p", obj).should == "<inspect-result>"
+      end
+    end
+
+    describe "s" do
+      it "substitute argument passes as a string" do
+        format("%s", "abc").should == "abc"
+      end
+
+      it "converts argument to string with to_s" do
+        obj = mock("string")
+        obj.should_receive(:to_s).and_return("abc")
+        format("%s", obj).should == "abc"
+      end
+
+      # strange behavior
+      # expected any core/std lib uses to_int/to_str/to_ary/to_hash for argument
+      it "does not try to convert with to_str" do
+        obj = BasicObject.new
+        def obj.to_str
+          "abc"
+        end
+
+        -> () {
+          format("%s", obj)
+        }.should raise_error(NoMethodError)
+      end
+    end
+
+    describe "%" do
+      it "displays percent sign itself" do
+        format("%").should == "%"
+      end
+
+      it "is escaped by %" do
+        format("%%").should == "%"
+        format("%%d", 10).should == "%d"
+      end
+    end
+  end
+
+  describe "flags" do
+    describe "space" do
+      context "applies to numeric formats bBdiouxXeEfgGaA" do
+        it "leaves a space at the start of non-negative numbers" do
+          format("% b", 10).should == " 1010"
+          format("% B", 10).should == " 1010"
+          format("% d", 112).should == " 112"
+          format("% i", 112).should == " 112"
+          format("% o", 87).should == " 127"
+          format("% u", 112).should == " 112"
+          format("% x", 196).should == " c4"
+          format("% X", 196).should == " C4"
+
+          format("% e", 109.52).should == " 1.095200e+02"
+          format("% E", 109.52).should == " 1.095200E+02"
+          format("% f", 10.952).should == " 10.952000"
+          format("% g", 12.1234).should == " 12.1234"
+          format("% G", 12.1234).should == " 12.1234"
+          format("% a", 196).should == " 0x1.88p+7"
+          format("% A", 196).should == " 0X1.88P+7"
+        end
+
+        it "does not leave a space at the start of negative numbers" do
+          format("% b", -10).should == "-1010"
+          format("% B", -10).should == "-1010"
+          format("% d", -112).should == "-112"
+          format("% i", -112).should == "-112"
+          format("% o", -87).should == "-127"
+          format("% u", -112).should == "-112"
+          format("% x", -196).should == "-c4"
+          format("% X", -196).should == "-C4"
+
+          format("% e", -109.52).should == "-1.095200e+02"
+          format("% E", -109.52).should == "-1.095200E+02"
+          format("% f", -10.952).should == "-10.952000"
+          format("% g", -12.1234).should == "-12.1234"
+          format("% G", -12.1234).should == "-12.1234"
+          format("% a", -196).should == "-0x1.88p+7"
+          format("% A", -196).should == "-0X1.88P+7"
+        end
+
+        # strange behavior
```

Comment:
Strange I agree, but documented it seems:
```
  Flag     | Applies to    | Meaning
  ---------+---------------+-----------------------------------------
  space    | bBdiouxX      | Leave a space at the start of
           | aAeEfgG       | non-negative numbers.
           | (numeric fmt) | For `o', `x', `X', `b' and `B', use
           |               | a minus sign with absolute value for
           |               | negative values.
```
Maybe specify it only applies for `o', `x', `X', `b' and `B' here?

Reply:
Agreed. Tests for negative values in cases other than 'o',x', X',b' and 'B' should be included.